File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        IEnumerable<int> Dfs(TreeNode node) {
            for (var stack = new Stack<TreeNode>();
                    node != null || stack.Count != 0; node = node.right) {
                // traverse left
                for (; node != null; node = node.left) stack.Push(node);

                // visit the subroot
                node = stack.Pop();
                yield return node.val;
            }
        }

        return Dfs(root).ToList();
    }
}

File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(node):
            stack = []
            while node or stack:
                while node: # traverse left
                    stack.append(node)
                    node = node.left

                node = stack.pop()
                yield node.val

                node = node.right

        return list(dfs(root))
